CONTRIBUTING.md

```diff
@@ -0,0 +1,437 @@
+    - In particular, use of self types should be avoided.
+- Avoid optionals in return values, fields and parameters.
+    - Where possible, you should prefer a `@Nullable T`.
+    - If vanilla exposes optionals in return types, then returning an optional is fine.
+- Avoid requiring the user to cast to a subtype if possible.
+    - Adding methods to vanilla types can be done via interface injection.
+
+### API design patterns to consider
+
+- Transitive Access Wideners (TAWs) should be used to expose access to private or protected members in vanilla classes.
+    - Most TAWs should go in the dedicated `fabric-transitive-access-wideners-v1` module.
+        - Remember to add the `transitive-` prefix, otherwise dependent mods will not see the access modifications.
+        - Some TAWs such as Block subclass constructors are automatically generated. Make sure that you don't edit the `.accesswidener` file in `src` directly. Rather edit the template file (`template.accesswidener`) and run the `gradlew generateAccessWidener` task to update the generated file.
+        - Large amounts of TAWs for a specific purpose can be included in another module, as is the case for the data generation API, for example.
+    - Do **not** expose TAWs for functions that take a `String` identifier.
+        - This makes it too easy to forget the mod ID namespace, so the identifier would often end up in the vanilla `minecraft` namespace.
+    - In general, keep the API guidelines in mind when deciding whether something should be a TAW.
+- Interface injection (i.e. making a minecraft class or interface extend a Fabric interface) should be considered over separate static helpers.
+    - Interface injection requires both a `fabric.mod.json` custom value to make it visible in Minecraft source code, and a mixin to actually implement the interface at runtime.
+    - Injected interfaces should have **no abstract methods**.
+        - Methods that are guaranteed to be implemented via a mixin to a vanilla class should contain a default body that throws an error. For example:
+        ```java
+        default void injectedMethod() {
+            throw new UnsupportedOperationException("Implemented via mixin");
+        }
+        ```
+        - Never use interface injection to add methods that modders must implement. Rather define a subclass or subinterface in Fabric API.
+- Builders can be used instead of constructors or factory methods with large amounts of parameters.
+
+### API class modifiers and member visibility
+
+- Classes in Fabric API should be `final` classes unless the class exposed in the API is explicitly meant for extension.
+- `private` constructors should be used in API classes unless the class is explicitly meant to be instantiated.
+    - This only applies to modder-facing classes, i.e. classes in the `net.fabricmc.fabric.api` subpackage. See below for package structure.
+    - They should be placed at the very bottom of the class to not hurt readability.
```

Comment:
Placing them at the very top also doesn't hurt readability imo (JDK code does this)